### PR TITLE
[SWY-122] Separate preview and production databases

### PR DESCRIPTION
**Pull request by justaddcl**
## Background
- Linear ticket: [SWY-122](https://linear.app/paranoid-android/issue/SWY-122/separate-preview-and-production-databases)
- Vercel Preview runs with production-style NODE_ENV, so relying on NODE_ENV for database selection can allow preview deployments to use production database credentials.
- This separates preview and production database selection while keeping local and E2E flows on DATABASE_URL.

## Summary by CodeRabbit

* **New Features**
  * Added clearer database environment handling for local, preview, E2E, and production setups.
  * Introduced preview-specific database scripts for push, seed, clear, and update workflows.

* **Bug Fixes**
  * Improved database URL selection so the app uses the correct connection string in each environment.
  * Added safeguards to catch missing or misconfigured database URLs earlier.

* **Documentation**
  * Expanded setup guidance to explain which environment variables to use in each deployment scenario.

- Adds a shared database URL resolver that chooses DATABASE_URL for local, E2E, and preview environments, POSTGRES_URL for production, and fails fast when preview is configured with the same URL as production.
- Wires the resolver into app database initialization and Drizzle config, with env validation updated for VERCEL_ENV, SANBI_APP_ENV, and SANBI_E2E.
- Adds preview database scripts and documents expected local, preview, and production database configuration.

## How to test
- Start the app locally with a normal DATABASE_URL and confirm the app loads against local data.
- Configure a preview deployment with a dedicated preview DATABASE_URL and confirm preview data is isolated from production.
- Confirm production deployments or production database scripts use POSTGRES_URL by setting SANBI_APP_ENV=production for local production script runs.

<h3>Greptile Summary</h3>

This PR replaces the `NODE_ENV === \"production\"` database selection heuristic (unsafe on Vercel Preview, which runs with a production-style `NODE_ENV`) with a dedicated `resolveDatabaseUrl` resolver that selects among local, E2E, preview, and production databases based on `SANBI_E2E`, `SANBI_APP_ENV`, and `VERCEL_ENV`. Both `DATABASE_URL` and `POSTGRES_URL` are now optional in `env.js`, with fail-fast validation deferred to the resolver.

- **New `resolveDatabaseUrl` resolver** (`src/server/db/resolveDatabaseUrl.ts`): priority-ordered environment detection with a safety check that throws when preview `DATABASE_URL` matches `POSTGRES_URL`.
- **Wired into both `db/index.ts` and `drizzle.config.ts`**: replaces the inline `NODE_ENV` ternary in both connection entry points.
- **New `db:*:preview` scripts** in `package.json` and documentation in `README.md` for operating against each database tier.

<h3>Confidence Score: 5/5</h3>

Safe to merge — the resolver is well-tested and the production-URL guard prevents preview deployments from accidentally connecting to the production database.

The core routing logic is correct and covered by thorough unit tests across all environment combinations. Both connection entry points (db/index.ts and drizzle.config.ts) are consistently updated. The fail-fast checks for missing URLs and the preview/production URL collision guard are solid safeguards.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/server/db/resolveDatabaseUrl.ts | New resolver with clear priority ordering (E2E → SANBI_APP_ENV → VERCEL_ENV → local fallback) and a preview URL equality guard; logic is correct and well-tested. |
| src/server/db/__tests__/resolveDatabaseUrl.test.ts | Comprehensive test coverage of all resolution paths including edge cases (E2E overrides production env, preview/production URL collision, missing URLs). |
| src/env.js | DATABASE_URL and POSTGRES_URL made optional; SANBI_APP_ENV, SANBI_E2E, and VERCEL_ENV added with correct enum constraints; emptyStringAsUndefined ensures blank values behave as absent. |
| src/server/db/index.ts | Replaces inline NODE_ENV ternary with resolveDatabaseUrl call; no behavioral change beyond the new routing logic. |
| drizzle.config.ts | Migrated from inline NODE_ENV check to resolveDatabaseUrl; consistent with the db/index.ts change. |
| package.json | Adds db:*:preview scripts mirroring the db:*:prod pattern; all prod scripts correctly updated from NODE_ENV=production to SANBI_APP_ENV=production. |
| .env.example | Documents that DATABASE_URL is for local/preview and POSTGRES_URL for production; adds SANBI_APP_ENV placeholder. |
| README.md | New Database Environments section clearly explains which URL and which scripts to use per environment tier. |

<h3>Flowchart</h3>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[resolveDatabaseUrl called] --> B{SANBI_E2E === '1'?}
    B -- Yes --> C[runtime = 'e2e']
    B -- No --> D{SANBI_APP_ENV set?}
    D -- development --> E[runtime = 'local']
    D -- preview/production --> F[runtime = SANBI_APP_ENV]
    D -- not set --> G{VERCEL_ENV set?}
    G -- preview/production --> H[runtime = VERCEL_ENV]
    G -- not set / development --> I[runtime = 'local']
    C --> J[require DATABASE_URL]
    E --> J
    I --> J
    F --> K{runtime = 'production'?}
    H --> K
    K -- Yes --> L[require POSTGRES_URL\nreturn POSTGRES_URL]
    K -- No --> M[require DATABASE_URL]
    J --> N[return DATABASE_URL]
    M --> O{runtime = 'preview' AND\nDATABASE_URL === POSTGRES_URL?}
    O -- Yes --> P[throw misconfiguration error]
    O -- No --> Q[return DATABASE_URL]
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
flowchart TD
    A[resolveDatabaseUrl called] --> B{SANBI_E2E === '1'?}
    B -- Yes --> C[runtime = 'e2e']
    B -- No --> D{SANBI_APP_ENV set?}
    D -- development --> E[runtime = 'local']
    D -- preview/production --> F[runtime = SANBI_APP_ENV]
    D -- not set --> G{VERCEL_ENV set?}
    G -- preview/production --> H[runtime = VERCEL_ENV]
    G -- not set / development --> I[runtime = 'local']
    C --> J[require DATABASE_URL]
    E --> J
    I --> J
    F --> K{runtime = 'production'?}
    H --> K
    K -- Yes --> L[require POSTGRES_URL\nreturn POSTGRES_URL]
    K -- No --> M[require DATABASE_URL]
    J --> N[return DATABASE_URL]
    M --> O{runtime = 'preview' AND\nDATABASE_URL === POSTGRES_URL?}
    O -- Yes --> P[throw misconfiguration error]
    O -- No --> Q[return DATABASE_URL]
```

</a>

<sub>Reviews (2): Last reviewed commit: ["Address database resolver review feedbac..."](https://github.com/justaddcl/sanbi/commit/44dee5b642a050cd1255b5ed9c6d1db372a50d9f) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=39821104)</sub>

### CODE REVIEW

File: src/server/db/resolveDatabaseUrl.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
+type DatabaseUrlEnv = {
+  DATABASE_URL?: string;
+  NODE_ENV?: string;
+  POSTGRES_URL?: string;
+  SANBI_APP_ENV?: string;
+  SANBI_E2E?: string;
+  VERCEL_ENV?: string;
+};
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/server/db/resolveDatabaseUrl.ts:7]: Addressed in https://github.com/justaddcl/sanbi/commit/44dee5b9a52ac4b4137ac4982d76f40f50c63f61 by removing the unused NODE_ENV field from the resolver input type.

File: src/server/db/resolveDatabaseUrl.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  env: DatabaseUrlEnv,
+): DatabaseRuntimeEnvironment => {
+  if (env.SANBI_E2E === "1") {
+    return "e2e";
+  }
+
+  // SANBI_APP_ENV is for local scripts and non-Vercel overrides; VERCEL_ENV is set by Vercel deployments.
+  if (env.SANBI_APP_ENV === "preview" || env.SANBI_APP_ENV === "production") {
+    return env.SANBI_APP_ENV;
+  }
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/server/db/resolveDatabaseUrl.ts:25]: Addressed in https://github.com/justaddcl/sanbi/commit/44dee5b9a52ac4b4137ac4982d76f40f50c63f61 by treating SANBI_APP_ENV=development as an explicit local override before checking VERCEL_ENV.

File: src/server/db/__tests__/resolveDatabaseUrl.test.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+  });
+
+  it("fails clearly when the selected database URL is missing", () => {
+    expect(() =>
+      resolveDatabaseUrl({
+        DATABASE_URL: previewDatabaseUrl,
+        VERCEL_ENV: "production",
+      }),
+    ).toThrow("POSTGRES_URL is required for the production database connection.");
+  });
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/server/db/__tests__/resolveDatabaseUrl.test.ts:128]: Addressed in https://github.com/justaddcl/sanbi/commit/44dee5b9a52ac4b4137ac4982d76f40f50c63f61 by adding missing-DATABASE_URL coverage for the local and E2E resolution paths.
